// @stopify/continuations-runtime/dist/src/runtime/abstractRuntime.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const assert = require("assert");
// We throw this exception when a continuation value is applied. i.e.,
// captureCC applies its argument to a function that throws this exception.
class Restore {
    constructor(stack, savedStack) {
        this.stack = stack;
        this.savedStack = savedStack;
    }
}
exports.Restore = Restore;
// We throw this exception to end the current turn and continue execution on the
// next turn.
class EndTurn {
    constructor(callback) {
        this.callback = callback;
    }
}
exports.EndTurn = EndTurn;
// This class is used by all the runtimes to start the stack capturing process.
class Capture {
    constructor(f, stack) {
        this.f = f;
        this.stack = stack;
    }
}
exports.Capture = Capture;
class RuntimeImpl {
    constructor(
    // Maximum number of frames that can be consumed by this runtime.
    stackSize, 
    // Number of frames to be restored from the captured stack onto the JS stack.
    restoreFrames, 
    // True when the instrumented program is capturing the stack.
    capturing = false) {
        this.stackSize = stackSize;
        this.restoreFrames = restoreFrames;
        this.capturing = capturing;
        /**
         *  A saved stack trace. This field is only used when a user-mode exception
         * is thrown.
         */
        this.stackTrace = [];
        if (isFinite(stackSize)) {
            assert(restoreFrames <= stackSize, 'Cannot restore more frames than stack size');
        }
        this.stack = [];
        this.savedStack = [];
        this.restoreFrames = restoreFrames;
        this.stackSize = stackSize;
        this.remainingStack = stackSize;
        this.mode = true;
        this.kind = undefined; // the worst
    }
    topK(f) {
        return {
            kind: 'top',
            f: () => {
                this.stack = [];
                this.mode = true;
                return f();
            },
            this: this,
        };
    }
    runtime(body, onDone) {
        while (true) {
            const result = this.abstractRun(body);
            if (result.type === 'normal' || result.type === 'exception') {
                if (this.savedStack.length > 0) {
                    const exception = result.type === 'exception';
                    const ss = this.savedStack;
                    const restarter = this.topK(() => {
                        if (exception) {
                            throw result.value;
                        }
                        else {
                            return result.value;
                        }
                    });
                    this.stack = [restarter];
                    for (let i = 0; i < this.restoreFrames && ss.length - 1 - i >= 0; i++) {
                        this.stack.push(ss.pop());
                    }
                    body = () => {
                        this.mode = false;
                        return this.stack[this.stack.length - 1].f();
                    };
                }
                else if (result.type === 'normal') {
                    assert(this.mode, 'execution completed in restore mode');
                    return onDone(result);
                }
                else if (result.type === 'exception') {
                    assert(this.mode, `execution completed in restore mode, error was: ${result.value}`);
                    const stack = this.stackTrace;
                    this.stackTrace = [];
                    return onDone({ type: 'exception', value: result.value, stack });
                }
            }
            else if (result.type === 'capture') {
                body = () => result.f.call(global, this.makeCont(result.stack));
            }
            else if (result.type === 'restore') {
                body = () => {
                    if (result.stack.length === 0) {
                        throw new Error(`Can't restore from empty stack`);
                    }
                    this.mode = false;
                    this.stack = result.stack;
                    this.savedStack = result.savedStack;
                    const frame = this.stack[this.stack.length - 1];
                    return this.stack[this.stack.length - 1].f.apply(frame.this || global, (frame.params || []));
                };
            }
            else if (result.type === 'end-turn') {
                return result.callback(onDone);
            }
        }
    }
    /** jumper.ts and captureLogics.ts insert calls to pushTrace. */
    pushTrace(line) {
        this.stackTrace.push(line);
    }
    /** jumper.ts inserts calls to clearTrace. */
    clearTrace() {
        this.stackTrace = [];
    }
}
exports.RuntimeImpl = RuntimeImpl;
