// @stopify/continuations-runtime/dist/src/runtime/lazyRuntime.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
const common = require("./abstractRuntime");
__export(require("./abstractRuntime"));
class LazyRuntime extends common.RuntimeImpl {
    constructor(stackSize, restoreFrames) {
        super(stackSize, restoreFrames);
        this.kind = 'lazy';
    }
    captureCC(f) {
        this.capturing = true;
        throw new common.Capture(f, []);
    }
    makeCont(stack) {
        const savedStack = this.savedStack;
        this.savedStack = [];
        for (let i = stack.length - 1; i >= this.restoreFrames; i -= 1) {
            savedStack.push(stack.pop());
        }
        return (x) => {
            let restarter = () => {
                if (x.type === 'exception') {
                    throw x.value;
                }
                else {
                    return x.value;
                }
            };
            throw new common.Restore([this.topK(restarter), ...stack], savedStack);
        };
    }
    endTurn(callback) {
        this.savedStack = [];
        this.stack = [];
        throw new common.EndTurn(callback);
    }
    abstractRun(body) {
        try {
            const v = body();
            return { type: 'normal', value: v };
        }
        catch (exn) {
            if (exn instanceof common.Capture) {
                this.capturing = false;
                return { type: 'capture', stack: exn.stack, f: exn.f };
            }
            else if (exn instanceof common.Restore) {
                return { type: 'restore', stack: exn.stack, savedStack: exn.savedStack };
            }
            else if (exn instanceof common.EndTurn) {
                return { type: 'end-turn', callback: exn.callback };
            }
            else {
                return { type: 'exception', value: exn };
            }
        }
    }
}
exports.LazyRuntime = LazyRuntime;
