// @stopify/estimators/dist/src/elapsedTimeEstimator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Interface for an object that estimates elapsed time.
 */
class ElapsedTimeEstimator {
    /**
     * Clean up any necessary state. Called from the runtime's `onEnd` function.
     */
    cancel() { }
}
exports.ElapsedTimeEstimator = ElapsedTimeEstimator;
class ExactTimeEstimator extends ElapsedTimeEstimator {
    constructor(last = Date.now()) {
        super();
        this.last = last;
    }
    reset() {
        this.last = Date.now();
    }
    elapsedTime() {
        return Date.now() - this.last;
    }
}
exports.ExactTimeEstimator = ExactTimeEstimator;
class CountdownTimeEstimator extends ElapsedTimeEstimator {
    constructor(timePerElapsed, i = 0) {
        super();
        this.timePerElapsed = timePerElapsed;
        this.i = i;
    }
    reset() {
        this.i = 0;
    }
    elapsedTime() {
        this.i++;
        return this.i * this.timePerElapsed;
    }
}
exports.CountdownTimeEstimator = CountdownTimeEstimator;
/** Draws a number from a geometric distribution. */
function geom(p) {
    return Math.ceil(Math.log(1 - Math.random()) / Math.log(1 - p));
}
class SampleAverageTimeEstimator extends ElapsedTimeEstimator {
    constructor(
    // total calls to elapsedTime
    i = 1, 
    // last value produced by Date.now()
    last = Date.now(), 
    // time between successive calls to elapsedTime
    timePerElapsed = 100, 
    // these many calls to elapsedTime between observations of time
    countDownFrom = geom(1 / i), 
    // countdown until we re-observe the time
    countDown = countDownFrom, 
    // number of times elapsedTime has been invoked since last reset
    elapsedTimeCounter = 0) {
        super();
        this.i = i;
        this.last = last;
        this.timePerElapsed = timePerElapsed;
        this.countDownFrom = countDownFrom;
        this.countDown = countDown;
        this.elapsedTimeCounter = elapsedTimeCounter;
    }
    elapsedTime() {
        this.i = (this.i + 1) | 0;
        this.elapsedTimeCounter = (this.elapsedTimeCounter + 1) | 0;
        if (this.countDown-- === 0) {
            const now = Date.now();
            this.timePerElapsed = (now - this.last) / this.countDownFrom;
            this.last = now;
            this.countDownFrom = geom(1 / this.i);
            this.countDown = this.countDownFrom;
        }
        return this.timePerElapsed * this.elapsedTimeCounter;
    }
    reset() {
        this.elapsedTimeCounter = 0;
    }
}
exports.SampleAverageTimeEstimator = SampleAverageTimeEstimator;
class VelocityEstimator extends ElapsedTimeEstimator {
    constructor(
    // Expected distance between resamples (units: time)
    resample, 
    // total calls to elapsedTime
    i = 1, 
    // Units: time
    lastPosition = Date.now(), 
    // Units: time / #elapsedTime
    velocityEstimate = 100000, 
    // Units: #elapsedTime;
    resampleTimespanEstimate = velocityEstimate, 
    // countdown until we re-observe the time
    countDown = 1, 
    // Distance since last reset. Units: #elapsedTime
    distance = 0) {
        super();
        this.resample = resample;
        this.i = i;
        this.lastPosition = lastPosition;
        this.velocityEstimate = velocityEstimate;
        this.resampleTimespanEstimate = resampleTimespanEstimate;
        this.countDown = countDown;
        this.distance = distance;
    }
    elapsedTime() {
        this.i = (this.i + 1) | 0;
        this.distance = (this.distance + 1) | 0;
        if (this.countDown-- === 0) {
            const currentPosition = Date.now();
            // NOTE(arjun): This is a small float. It may be a good idea to scale
            // everything up to an integer.
            this.velocityEstimate = this.resampleTimespanEstimate / (currentPosition - this.lastPosition);
            this.lastPosition = currentPosition;
            this.resampleTimespanEstimate = Math.max((this.resample * this.velocityEstimate) | 0, 10);
            this.countDown = this.resampleTimespanEstimate;
        }
        return (this.distance / this.velocityEstimate) | 0;
    }
    reset() {
        this.distance = 0;
    }
}
exports.VelocityEstimator = VelocityEstimator;
