// @stopify/continuations/dist/src/runtime/implicitApps.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function toPrimitive(x) {
    // TODO(arjun): This isn't exactly the right order.
    if (typeof x === 'object' && x !== 'null') {
        let v = x.valueOf();
        if (typeof v === 'object' && v !== 'null') {
            v = x.toString();
            if (typeof v === 'object' && v !== 'null') {
                return undefined;
            }
            return v;
        }
        return v;
    }
    return x;
}
function toKey(x) {
    const r = toPrimitive(x);
    if (typeof r === 'string' || typeof r === 'number') {
        return r;
    }
    else {
        return String(r);
    }
}
exports.toKey = toKey;
function add(x, y) {
    return toPrimitive(x) + toPrimitive(y);
}
exports.add = add;
function mul(x, y) {
    return toPrimitive(x) * toPrimitive(y);
}
exports.mul = mul;
function sub(x, y) {
    return toPrimitive(x) - toPrimitive(y);
}
exports.sub = sub;
function div(x, y) {
    return toPrimitive(x) / toPrimitive(y);
}
exports.div = div;
