// @stopify/continuations/dist/src/callcc/declVars.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Assumes that singleVarDecls has run before this and handled all the `let`
 * and `const` declarations.
 *
 * Moves var statements to the top of functions.
 *
 * When a var is moved, it is initialized to undefined. The
 * transformation introduces assignment statements where the var statement
 * originally occurred.
 */
const t = require("babel-types");
const h = require("@stopify/util");
const funWithBody = {
    enter(path) {
        path.node.decls = [];
    },
    exit(path) {
        if (path.node.decls.length > 0) {
            const decl = t.variableDeclaration('var', path.node.decls.map(id => t.variableDeclarator(id)));
            path.node.body.body.unshift(decl);
        }
    }
};
const lift = {
    Program: {
        enter(path) {
            path.node.decls = [];
        },
        exit(path) {
            if (path.node.decls.length > 0) {
                const decl = t.variableDeclaration('var', path.node.decls.map(id => t.variableDeclarator(id)));
                path.node.body.unshift(decl);
            }
        },
    },
    FunctionDeclaration: funWithBody,
    FunctionExpression: funWithBody,
    ObjectMethod: funWithBody,
    VariableDeclaration(path) {
        const { declarations } = path.node;
        let fParent = path.getFunctionParent().node;
        if (!h.isFunWithBody(fParent) && !t.isProgram(fParent)) {
            throw new Error(`Variable declarations should be inside a function. Parent was ${path.node.type}`);
        }
        const stmts = [];
        if (declarations[0].__boxVarsInit__) {
            return;
        }
        for (const decl of declarations) {
            if (decl.id.type !== 'Identifier') {
                throw new Error(`Destructuring assignment not supported`);
            }
            fParent.decls.push(decl.id);
            if (decl.init !== null) {
                // If we call path.insertAfter here, we will add assignments in reverse
                // order. Fortunately, path.replaceWithMultiple can take an array of nodes.
                stmts.push(t.expressionStatement(t.assignmentExpression('=', decl.id, decl.init)));
            }
        }
        path.replaceWithMultiple(stmts);
    }
};
module.exports = function () {
    return { visitor: lift };
};
