// @stopify/continuations/dist/src/callcc/label.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const t = require("babel-types");
const cannotCapture_1 = require("../common/cannotCapture");
const h = require("@stopify/util");
const imm = require("immutable");
var AppType;
(function (AppType) {
    AppType[AppType["None"] = 0] = "None";
    AppType[AppType["Tail"] = 1] = "Tail";
    AppType[AppType["Mixed"] = 2] = "Mixed";
})(AppType = exports.AppType || (exports.AppType = {}));
function joinAppType(x, y) {
    return Math.max(x, y);
}
function joinAppTypes(...args) {
    return args.reduce(joinAppType, AppType.None);
}
function getLabels0(node) {
    if (node === null) {
        return [];
    }
    return node.labels === undefined ? [] : [...node.labels];
}
function getLabels(...nodes) {
    const r = [];
    for (const node of nodes) {
        r.push(...getLabels0(node));
    }
    return [...(new Set(r)).values()];
}
exports.getLabels = getLabels;
function getAppType(node) {
    if (node === null) {
        return AppType.None;
    }
    return node.appType;
}
// true if the expression is a function call that may capture a continuation
function isUnsafeCall(e) {
    return ((t.isCallExpression(e) || t.isNewExpression(e)) && !cannotCapture_1.cannotCapture(e));
}
const visitFunction = {
    enter(path) {
        this.inTryBlockStack.push(this.inTryBlock);
        this.inTryBlock = false;
        const allVars = imm.Set.of(...Object.keys(path.scope.bindings));
        let otherVars = imm.Set.of(...path.node.params.map(h.lvaltoName))
            .union(path.node.id ? imm.Set.of(path.node.id.name) : imm.Set());
        path.node.localVars = allVars.subtract(otherVars).toArray()
            .map(x => t.identifier(x));
    },
    exit(path) {
        this.inTryBlock = this.inTryBlockStack.pop();
        if (path.node.type === 'FunctionDeclaration') {
            path.node.appType = AppType.None;
        }
    }
};
const visitor = {
    TryStatement: {
        enter(path) {
            this.inTryBlockStack.push(this.inTryBlock);
            this.inTryBlock = true;
        },
        exit(path) {
            const { block, finalizer, handler } = path.node;
            this.inTryBlock = this.inTryBlockStack.pop();
            path.node.labels = getLabels(block, finalizer, handler && handler.body);
            // TODO(arjun): This is not exactly right. A try statement has tail calls
            // if and only if the try block has no function calls and all function
            // calls in the catch and finally block are tail calls.
            const t = joinAppTypes(getAppType(block), getAppType(finalizer), getAppType(handler && handler.body));
            path.node.appType = (t === AppType.None) ? AppType.None : AppType.Mixed;
        }
    },
    // TODO(arjun): I think an enclosing function will gather labels from
    // an enclosed function.
    FunctionDeclaration: visitFunction,
    FunctionExpression: visitFunction,
    ObjectMethod: visitFunction,
    Program: {
        enter(path) {
            this.counter = 0;
            this.inTryBlock = false;
            this.inTryBlockStack = [];
        }
    },
    ReturnStatement: {
        exit(path) {
            // Assumes no nested calls.
            const isCall = isUnsafeCall(path.node.argument);
            if (!isCall) {
                path.node.appType = AppType.None;
            }
            else if (this.inTryBlock) {
                path.node.appType = AppType.Mixed;
            }
            else {
                path.node.appType = AppType.Tail;
            }
        }
    },
    CallExpression: function (path) {
        path.node.labels = [this.counter++];
    },
    NewExpression: function (path) {
        path.node.labels = [this.counter++];
    },
    AssignmentExpression: {
        exit(path) {
            path.node.labels = getLabels(path.node.right);
        }
    },
    VariableDeclaration: {
        exit(path) {
            path.node.appType = AppType.None;
        }
    },
    ExpressionStatement: {
        exit(path) {
            const unsafe = t.isAssignmentExpression(path.node.expression) &&
                isUnsafeCall(path.node.expression.right);
            path.node.labels = getLabels(path.node.expression);
            path.node.appType = unsafe ? AppType.Mixed : AppType.None;
        }
    },
    IfStatement: {
        exit(path) {
            const { consequent, alternate } = path.node;
            path.node.labels = getLabels(consequent, alternate);
            path.node.appType = joinAppTypes(getAppType(consequent), getAppType(alternate));
        }
    },
    WhileStatement: {
        exit(path) {
            const { body } = path.node;
            path.node.labels = getLabels(body);
            path.node.appType = getAppType(body);
        }
    },
    LabeledStatement: {
        exit(path) {
            const { body } = path.node;
            path.node.labels = getLabels(body);
            path.node.appType = getAppType(body);
        }
    },
    BlockStatement: {
        exit(path) {
            const { body } = path.node;
            path.node.labels = getLabels(...body);
            path.node.appType = joinAppTypes(...body.map(getAppType));
        }
    },
};
function plugin() {
    return { visitor };
}
exports.plugin = plugin;
