// @stopify/stopify/dist/src/compiler/compiler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const babel = require("babel-core");
const stopifyCallCC = require("../stopify/stopifyCallCC");
const babylon = require("babylon");
/**
 * Compiles (i.e., "stopifies") a program. This function should not be used
 * directly by clients of Stopify.
 *
 * @param src the program to Stopify
 * @param opts compiler options
 * @returns the stopified program
 */
function compileFromAst(src, opts) {
    const babelOpts = {
        plugins: [[stopifyCallCC.plugin, opts]],
        babelrc: false,
        ast: false,
        code: true,
        minified: true,
        comments: false,
    };
    const { code } = babel.transformFromAst(src, undefined, babelOpts);
    return code;
}
exports.compileFromAst = compileFromAst;
/**
 * Compiles (i.e., "stopifies") a program. This function should not be used
 * directly by clients of Stopify.
 *
 * @param src the program to Stopify
 * @param opts compiler options
 * @returns the stopified program
 */
function compile(src, opts) {
    return compileFromAst(babylon.parse(src).program, opts);
}
exports.compile = compile;
