// @stopify/util/dist/ts/babelHelpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const t = require("babel-types");
function isFunWithBody(node) {
    return t.isFunctionDeclaration(node) || t.isFunctionExpression(node) ||
        t.isObjectMethod(node);
}
exports.isFunWithBody = isFunWithBody;
exports.eTrue = t.booleanLiteral(true);
exports.eFalse = t.booleanLiteral(false);
exports.eUndefined = t.identifier('undefined');
function enclosingScopeBlock(path) {
    const parent = path.getFunctionParent().node;
    if (t.isProgram(parent)) {
        return parent.body;
    }
    else if (t.isFunctionExpression(parent) ||
        t.isFunctionDeclaration(parent) ||
        t.isObjectMethod(parent)) {
        return parent.body.body;
    }
    else {
        throw new Error(`parent is a ${parent.type}`);
    }
}
exports.enclosingScopeBlock = enclosingScopeBlock;
/** Constructs an 'e1 && e2', but simplifies when either sub-expression is
 *  a literal.
 */
function and(e1, e2) {
    if ((e1.type === 'BooleanLiteral' && e1.value === true)) {
        return e2;
    }
    else if (e2.type === 'BooleanLiteral' && e2.value === true) {
        return e1;
    }
    else if ((e1.type === 'BooleanLiteral' && e1.value === false) ||
        (e2.type === 'BooleanLiteral' && e2.value === false)) {
        return exports.eFalse;
    }
    else {
        return t.logicalExpression('&&', e1, e2);
    }
}
exports.and = and;
/** Constructs an 'e1 || e2', but simplifies when either sub-expression is
 *  a literal.
 */
function or(...es) {
    let r = t.booleanLiteral(false);
    for (const e of es) {
        r = orBin(r, e);
    }
    return r;
}
exports.or = or;
function orBin(e1, e2) {
    if ((e1.type === 'BooleanLiteral' && e1.value === false)) {
        return e2;
    }
    else if (e2.type === 'BooleanLiteral' && e2.value === false) {
        return e1;
    }
    else if ((e1.type === 'BooleanLiteral' && e1.value === true) ||
        (e2.type === 'BooleanLiteral' && e2.value === true)) {
        return exports.eTrue;
    }
    else {
        return t.logicalExpression('||', e1, e2);
    }
}
function sIf(e, s1, s2) {
    if (s2 === undefined) {
        s2 = t.emptyStatement();
    }
    if (e.type === 'BooleanLiteral' && e.value === true) {
        return s1;
    }
    else if (e.type === 'BooleanLiteral' && e.value === false) {
        return s2;
    }
    else {
        if (t.isEmptyStatement(s2)) {
            return t.ifStatement(e, s1);
        }
        else {
            return t.ifStatement(e, s1, s2);
        }
    }
}
exports.sIf = sIf;
/**
 * Replaces a statement with a sequence of statements, creating a BlockStatement
 * if necessary.
 *
 * NOTE(arjun): There appears to be a bug with Babel's path.replaceWithMultiple
 * that this function works around. To witness the bug, try the ClojureScript
 * benchmarks using Babel's replaceWithMultiple instead of this function.
 *
 * @param path the path to a statement to replace
 * @param stmts a sequence of statements
 */
function replaceWithStatements(path, ...stmts) {
    if (path.parent.type === 'BlockStatement') {
        path.replaceWithMultiple(stmts);
    }
    else {
        path.replaceWith(t.blockStatement(stmts));
    }
}
exports.replaceWithStatements = replaceWithStatements;
/**
 * Given an 'LVal' that is an identifier, produces the identifier's name.
 * Throws an exception if the 'LVal' is not an identifier.
 *
 * @param lval an l-value
 * @returns the name of the identifier, if 'lval' is an identifier
 */
function lvaltoName(lval) {
    if (lval.type === 'Identifier') {
        return lval.name;
    }
    else if (lval.type === 'RestElement' && lval.argument.type === 'Identifier') {
        return lval.argument.name;
    }
    else {
        throw new Error(`Expected Identifier, received ${lval.type}`);
    }
}
exports.lvaltoName = lvaltoName;
function isPropertyValue(p) {
    return (t.isObjectMethod(p) ||
        (t.isObjectProperty(p) &&
            p.computed === false &&
            t.isExpression(p.value) &&
            isValue(p.value)));
}
/**
 * Produces 'true' if 'e' is a value.
 *
 * @param e
 */
function isValue(e) {
    if (e === null) {
        return false;
    }
    if (t.isLiteral(e) || t.isFunction(e) || t.isIdentifier(e)) {
        return true;
    }
    if (t.isArrayExpression(e)) {
        return e.elements.every(isValue);
    }
    if (t.isObjectExpression(e)) {
        return e.properties.every(isPropertyValue);
    }
    return false;
}
exports.isValue = isValue;
function arrayPrototypeSliceCall(e) {
    return t.callExpression(t.memberExpression(t.memberExpression(t.memberExpression(t.identifier('Array'), t.identifier('prototype')), t.identifier('slice')), t.identifier('call')), [e]);
}
exports.arrayPrototypeSliceCall = arrayPrototypeSliceCall;
function varDecl(x, init) {
    const id = typeof x === 'string' ? t.identifier(x) : x;
    return t.variableDeclaration('var', [t.variableDeclarator(id, init)]);
}
exports.varDecl = varDecl;
function enclosingFunctionName(path) {
    // TODO(arjun): this traversal is slow
    const f = path.getFunctionParent().node;
    if (t.isFunctionExpression(f)) {
        return f.originalName || f.id.name;
    }
    else if (t.isFunctionDeclaration(f)) {
        return f.id.name;
    }
    else {
        return;
    }
}
exports.enclosingFunctionName = enclosingFunctionName;
function returnLast(statements) {
    const N = statements.length - 1;
    const last = statements[N];
    if (t.isExpressionStatement(last)) {
        statements[N] = t.returnStatement(last.expression);
    }
    return statements;
}
exports.returnLast = returnLast;
function letExpression(name, value, kind = 'let') {
    return t.variableDeclaration(kind, [t.variableDeclarator(name, value)]);
}
exports.letExpression = letExpression;
