// @stopify/util/dist/ts/helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const babel = require("babel-core");
const t = require("babel-types");
var babelHelpers_1 = require("./babelHelpers");
exports.letExpression = babelHelpers_1.letExpression;
// Helper to generate tagging function for AST tags preserved between traversals.
function tag(tag, t, v) {
    const tagged = t;
    tagged[tag] = v;
    return tagged;
}
exports.tag = tag;
exports.breakLbl = (t, v) => tag('break_label', t, v);
exports.continueLbl = (t, v) => tag('continue_label', t, v);
exports.newTag = (t) => tag('new', t, true);
const containsCallVisitor = {
    FunctionExpression(path) {
        path.skip();
    },
    CallExpression(path) {
        this.containsCall = true;
        path.stop();
    },
    NewExpression(path) {
        this.containsCall = true;
        path.stop();
    },
};
/**
 * Traverses children of `path` and returns true if it contains any applications.
 */
function containsCall(path) {
    let o = { containsCall: false };
    path.traverse(containsCallVisitor, o);
    return o.containsCall;
}
exports.containsCall = containsCall;
/**
 * Use this when the contents of the body need to be flattened.
 * @param body An array of statements
 * @returns a new block (does not update the argument)
 */
function flatBodyStatement(body) {
    const newBody = [];
    body.forEach((elem) => {
        if (t.isBlockStatement(elem)) {
            elem.body.forEach((e) => {
                if (t.isStatement(e)) {
                    newBody.push(e);
                }
                else if (t.isEmptyStatement(e)) { }
                else {
                    throw new Error('Could not flatten body, element was not a statement');
                }
            });
        }
        else {
            newBody.push(elem);
        }
    });
    return t.blockStatement(newBody);
}
exports.flatBodyStatement = flatBodyStatement;
/**
 * A simple wrapper around Babel's `transformFromAst` function.
 */
function transformFromAst(path, plugins, ast = false, code = false) {
    const opts = {
        plugins: plugins,
        babelrc: false,
        code: false,
        ast: false,
    };
    return babel.transformFromAst(path.node, undefined, opts);
}
exports.transformFromAst = transformFromAst;
