// @stopify/normalize-js/dist/ts/anf.js
"use strict";
/**
 * Plugin to transform JS programs into ANF form.
 *
 * WARNING:
 * The plugin assumes that the assumptions stated in ./src/desugarLoop.js
 * hold. The resulting output is not guarenteed to be in ANF form if the
 * assumptions do not hold.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const t = require("babel-types");
const h = require("@stopify/util");
const hygiene_1 = require("@stopify/hygiene");
function withinTryBlock(path) {
    const funOrTryParent = path.findParent(p => p.isFunction() || p.isTryStatement());
    return t.isTryStatement(funOrTryParent);
}
const anfVisitor = {
    FunctionExpression: function (path) {
        const p = path.parent;
        if (!t.isVariableDeclarator(p)) {
            // Name the function application if it is not already named.
            const name = hygiene_1.fresh('fun');
            const bind = h.letExpression(name, path.node);
            path.getStatementParent().insertBefore(bind);
            path.replaceWith(name);
        }
    },
    ArrayExpression: function (path) {
        if (!h.containsCall(path)) {
            return;
        }
        const { elements: es } = path.node;
        let elements = es;
        elements.forEach((e, i) => {
            const id = hygiene_1.fresh('element');
            path.getStatementParent().insertBefore(h.letExpression(id, e));
            path.node.elements[i] = id;
        });
    },
    ObjectExpression: function (path) {
        if (!h.containsCall(path)) {
            return;
        }
        const { properties } = path.node;
        properties.forEach((p2, i) => {
            let p = p2;
            // TODO(arjun): Enable this code to expose a bug.
            // if (!t.isObjectProperty(p)) {
            //   throw new Error(`Expected ObjectProperty but got ${p.type}`);
            // }
            const id = hygiene_1.fresh('element');
            path.getStatementParent().insertBefore(h.letExpression(id, p.value));
            path.node.properties[i].value = id;
        });
    },
    CallExpression: {
        enter(path) {
            if (h.containsCall(path)) {
                if (t.isCallExpression(path.node.callee)) {
                    const id = hygiene_1.fresh('callee');
                    path.getStatementParent().insertBefore(h.letExpression(id, path.node.callee));
                    path.node.callee = id;
                }
                path.node.arguments.forEach((e, i) => {
                    if (!t.isExpression(e)) {
                        throw new Error(`Expected expression`);
                    }
                    const id = hygiene_1.fresh('arg');
                    path.getStatementParent().insertBefore(h.letExpression(id, e));
                    path.node.arguments[i] = id;
                });
            }
        },
        exit(path, state) {
            if (path.node.callee.mark === 'Flat') {
                return;
            }
            const p = path.parent;
            if ((!t.isVariableDeclarator(p) &&
                (!t.isReturnStatement(p) || state.opts.nameReturns)) ||
                (t.isReturnStatement(p) &&
                    withinTryBlock(path))) {
                // Name the function application if it is not already named or
                // if it is not a flat application.
                const name = hygiene_1.fresh('app');
                const bind = h.letExpression(name, path.node);
                path.getStatementParent().insertBefore(bind);
                if (path.parent.type === 'ExpressionStatement') {
                    path.remove();
                }
                else {
                    path.replaceWith(name);
                }
            }
        }
    },
    NewExpression: function (path) {
        if (path.node.callee.mark === 'Flat') {
            return;
        }
        const p = path.parent;
        if (!t.isVariableDeclarator(p)) {
            // Name the function application if it is not already named.
            const name = hygiene_1.fresh('app');
            const bind = h.letExpression(name, path.node);
            path.getStatementParent().insertBefore(bind);
            path.replaceWith(name);
        }
    },
};
module.exports = function () {
    return { visitor: anfVisitor };
};
