// @stopify/normalize-js/dist/ts/desugarSwitch.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const t = require("babel-types");
const util_1 = require("@stopify/util");
function ifTest(e, discriminant, fallthrough) {
    return t.logicalExpression('||', t.binaryExpression('===', discriminant, e), fallthrough);
}
function desugarCases(cases, discriminant, fallthrough) {
    if (cases.length === 0) {
        return [];
    }
    const [hd, ...tail] = cases;
    const { test, consequent } = hd;
    if (test === null) {
        return consequent;
    }
    else {
        const cond = ifTest(test, discriminant, fallthrough);
        const newTail = desugarCases(tail, discriminant, fallthrough);
        newTail.unshift(t.ifStatement(cond, t.blockStatement([
            t.expressionStatement(t.assignmentExpression('=', fallthrough, t.booleanLiteral(true))),
            ...consequent
        ])));
        return newTail;
    }
}
const switchVisitor = {
    BreakStatement: function (path) {
        const label = path.node.label;
        if (label === null) {
            const labeledParent = path.findParent(p => p.isLoop() || p.isSwitchStatement());
            if (labeledParent === null) {
                return;
            }
            path.node.label = labeledParent.node.break_label;
        }
    },
    SwitchStatement: {
        enter(path) {
            if (t.isLabeledStatement(path.parent)) {
                return;
            }
            const breakLabel = path.scope.generateUidIdentifier('switch');
            path.node = util_1.breakLbl(path.node, breakLabel);
            const labeledStatement = t.labeledStatement(breakLabel, path.node);
            path.replaceWith(labeledStatement);
        },
        exit(path) {
            const { discriminant, cases } = path.node;
            const test = path.scope.generateUidIdentifier('test');
            const fallthrough = path.scope.generateUidIdentifier('fallthrough');
            let desugared = [];
            desugared.unshift(...desugarCases(cases, test, fallthrough));
            desugared.unshift(util_1.letExpression(fallthrough, t.booleanLiteral(false)));
            desugared.unshift(util_1.letExpression(test, discriminant));
            path.replaceWith(t.blockStatement(desugared));
        }
    }
};
module.exports = function () {
    return { visitor: switchVisitor };
};
