// @stopify/normalize-js/dist/ts/freeIds.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const babel = require("babel-core");
const SetExt = require("./setExt");
const hasNff = ["FunctionDeclaration", "FunctionExpression", "Program"];
const functionTypes = ['FunctionDeclaration', 'FunctionExpression'];
const visitor = {
    Scope: {
        enter(path) {
            path.node.nestedFunctionFree = new Set();
            this.refIdStack.push(this.refIds);
            this.refIds = new Set();
        },
        exit(path) {
            const boundIds = new Set(Object.keys(path.scope.bindings));
            let freeIds = SetExt.diff(this.refIds, boundIds);
            path.node.freeIds = freeIds;
            this.refIds = this.refIdStack.pop();
            for (const x of freeIds) {
                this.refIds.add(x);
            }
            if (functionTypes.includes(path.node.type)) {
                const parent = enclosingFunction(path);
                const nestedFunctionFree = parent.node.nestedFunctionFree;
                for (const x of path.node.freeIds) {
                    nestedFunctionFree.add(x);
                }
            }
        }
    },
    ReferencedIdentifier(path) {
        const parentType = path.parent.type;
        if (parentType === "BreakStatement" ||
            parentType === 'ContinueStatement' ||
            parentType === "LabeledStatement") {
            return;
        }
        this.refIds.add(path.node.name);
    },
    BindingIdentifier(path) {
        const parentType = path.parent.type;
        if (parentType !== "AssignmentExpression") {
            return;
        }
        this.refIds.add(path.node.name);
    },
    Program(path) {
        path.node.nestedFunctionFree = new Set();
        this.refIdStack = [];
        this.refIds = new Set();
    }
};
function annotate(path) {
    const opts = {
        plugins: [[() => ({ visitor })]],
        babelrc: false,
        code: false,
        ast: false
    };
    babel.transformFromAst(path.node, undefined, opts);
}
exports.annotate = annotate;
function enclosingFunction(path) {
    const p = path.findParent(p => hasNff.includes(p.node.type));
    return p;
}
exports.enclosingFunction = enclosingFunction;
function isNestedFree(path, x) {
    return path.node.nestedFunctionFree.has(x);
}
exports.isNestedFree = isNestedFree;
function main() {
    const filename = process.argv[2];
    const opts = { plugins: [() => ({ visitor })], babelrc: false };
    babel.transformFile(filename, opts, (err, result) => {
        if (err !== null) {
            throw err;
        }
        console.log(result.code);
    });
}
if (require.main === module) {
    main();
}
