// @stopify/normalize-js/dist/ts/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const babel = require("babel-core");
const callcc = require("./callcc");
const freeIds = require("./freeIds");
exports.freeIds = freeIds;
const generic = require("./generic");
exports.generic = generic;
const babylon = require("babylon");
var callcc_1 = require("./callcc");
exports.plugin = callcc_1.default;
var generic_1 = require("./generic");
exports.unreachable = generic_1.unreachable;
const hygiene = require("@stopify/hygiene");
const h = require("@stopify/util");
const visitor = {
    Program(path) {
        h.transformFromAst(path, [[hygiene.plugin, { reserved: [] }]]);
        h.transformFromAst(path, [callcc.default]);
        path.stop();
    }
};
function normalize(code) {
    let ast = babylon.parse(code);
    const result = babel.transformFromAst(ast.program, undefined, {
        babelrc: false,
        code: true,
        ast: false,
        plugins: [() => ({ visitor })]
    });
    return result.code;
}
exports.normalize = normalize;
