// @stopify/normalize-js/dist/ts/jumperizeTry.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const t = require("babel-types");
const bh = require("@stopify/util");
const hygiene_1 = require("@stopify/hygiene");
const returnSentinal = t.identifier('finally_rv');
const throwSentinal = t.identifier('finally_exn');
const sentinal = t.memberExpression(t.identifier('$__C'), t.identifier('RV_SENTINAL'));
const sentinalExn = t.memberExpression(t.identifier('$__C'), t.identifier('EXN_SENTINAL'));
const visitor = {
    TryStatement: {
        enter(path) {
            if (path.node.finalizer) {
                this.inTryBlockStack.push(this.inTryWithFinally);
                this.inTryWithFinally = true;
            }
            if (path.node.handler) {
                const x = hygiene_1.fresh('e');
                path.insertBefore(t.variableDeclaration('var', [t.variableDeclarator(x)]));
                path.node.handler.eVar = x;
                path.node.handler.body.body.unshift(t.expressionStatement(t.assignmentExpression('=', x, path.node.handler.param)));
                if (path.node.finalizer) {
                    path.node.handler.body.body.unshift(t.expressionStatement(t.assignmentExpression('=', throwSentinal, sentinalExn)));
                }
            }
        },
        exit(path) {
            if (path.node.finalizer) {
                // NOTE(arjun): If we have several finally blocks in the same scope,
                // this probably creates duplicate declarations.
                const sentinalDecl0 = t.variableDeclaration('let', [
                    t.variableDeclarator(returnSentinal, sentinal)
                ]);
                const sentinalDecl1 = t.variableDeclaration('let', [
                    t.variableDeclarator(throwSentinal, sentinalExn),
                ]);
                bh.enclosingScopeBlock(path).unshift(sentinalDecl0);
                bh.enclosingScopeBlock(path).unshift(sentinalDecl1);
                path.node.finalizer.body.push(bh.sIf(t.binaryExpression('!==', returnSentinal, sentinal), t.returnStatement(returnSentinal), bh.sIf(t.binaryExpression('!==', throwSentinal, sentinalExn), t.throwStatement(throwSentinal))));
            }
            this.inTryWithFinally = this.inTryBlockStack.pop();
        }
    },
    ReturnStatement: {
        exit(path) {
            if (this.inTryWithFinally) {
                const arg = path.node.argument ||
                    t.unaryExpression('void', t.numericLiteral(0));
                path.insertBefore(t.expressionStatement(t.assignmentExpression('=', returnSentinal, arg)));
                path.replaceWith(t.returnStatement(returnSentinal));
                path.skip();
            }
        }
    },
    ThrowStatement: {
        exit(path) {
            if (this.inTryWithFinally) {
                const arg = path.node.argument ||
                    t.unaryExpression('void', t.numericLiteral(0));
                path.insertBefore(t.expressionStatement(t.assignmentExpression('=', throwSentinal, arg)));
                path.replaceWith(t.throwStatement(throwSentinal));
                path.skip();
            }
        }
    },
    Function: {
        enter(path) {
            this.inTryBlockStack.push(this.inTryWithFinally);
            this.inTryWithFinally = false;
        },
        exit(path) {
            this.inTryWithFinally = this.inTryBlockStack.pop();
        }
    },
    Program: {
        enter(path) {
            this.inTryWithFinally = false;
            this.inTryBlockStack = [];
        }
    },
};
function default_1() {
    return { visitor };
}
exports.default = default_1;
