// @stopify/hygiene/dist/ts/fastFreshId.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const util_1 = require("@stopify/util");
const t = require("babel-types");
const assert = require("assert");
const trie = require('trie'); // NOTE(arjun): No @types on 08/28/2017
let isInitialized = false;
// We populate this trie will all the identifiers in the program.
let prefixes = new trie.Trie();
const known = new Map();
// NOTE(arjun): the Visitor type does not have ReferencedIdentifier
// and BindingIdentifier on 08/28/2017.
const visitor = {
    ReferencedIdentifier(path) {
        prefixes.addWord(path.node.name);
    },
    BindingIdentifier(path) {
        prefixes.addWord(path.node.name);
    }
};
function plugin() {
    return { visitor: visitor };
}
exports.plugin = plugin;
function init(path) {
    known.clear();
    util_1.transformFromAst(path, [plugin]);
    isInitialized = true;
}
exports.init = init;
function reset() {
    prefixes = new trie.Trie();
    isInitialized = false;
}
exports.reset = reset;
function fresh(base) {
    assert(isInitialized, 'init() must be applied before fresh()');
    const k = known.get(base);
    if (typeof k !== 'undefined') {
        const x = k.newBase + String(k.i);
        k.i = k.i + 1;
        return t.identifier(x);
    }
    let j = 0;
    let newBase = base;
    while (prefixes.isValidPrefix(newBase)) {
        newBase = newBase + String(j);
        j = j + 1;
    }
    known.set(base, { newBase: newBase, i: 1 });
    return t.identifier(newBase + String(0));
}
exports.fresh = fresh;
function nameExprBefore(path, expr, base) {
    if (t.isIdentifier(expr)) {
        return expr;
    }
    const x = fresh(base || "x");
    path.insertBefore(t.variableDeclaration('let', [t.variableDeclarator(x, expr)]));
    return x;
}
exports.nameExprBefore = nameExprBefore;
