// @stopify/continuations/dist/src/compiler/markFlatFunctions.js
"use strict";
/**
 * Mark functions are not provably flat as 'Untransformed' and others as
 * 'Transformed'. This only marks the approproiate AST nodes. It is the job of
 * the transformation to use the information.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const t = require("babel-types");
let debug = false;
const prog = {
    // Mark the untransformed functions after running call expression markers.
    exit(path) {
        path.traverse(markUntransformed);
    }
};
const funcMarkUntransformed = {
    enter(path) {
        if (!path.node.mark) {
            path.node.mark = 'Flat';
        }
        if (debug && t.isFunctionDeclaration(path.node)) {
            const { loc, mark } = path.node;
            const info = `${mark}: ${path.node.id.name}${loc ? ` on line ${loc.start.line}` : ''}`;
            console.error(info);
        }
        if (debug && t.isFunctionExpression(path.node)) {
            const { loc, mark, id } = path.node;
            const info = `${mark}: ${id ? id.name : '<Anonymous>'}` +
                `${loc ? ` on line ${loc.start.line}` : ''}`;
            console.error(info);
        }
    }
};
const markUntransformed = {
    FunctionDeclaration: funcMarkUntransformed,
    FunctionExpression: funcMarkUntransformed,
};
const callExpr = {
    enter(path) {
        if (path.node.mark === 'Flat') {
            return;
        }
        const fParent = path.findParent(p => t.isFunctionDeclaration(p) || t.isFunctionExpression(p));
        if (fParent !== null && !fParent.node.mark) {
            fParent.node.mark = 'NotFlat';
        }
    }
};
const visitor = {
    Program: prog,
    CallExpression: callExpr,
    NewExpression: callExpr,
    Loop: callExpr,
    ThrowStatement: callExpr
};
function markFlatFunctions() {
    return { visitor };
}
exports.markFlatFunctions = markFlatFunctions;
