// @stopify/continuations/dist/src/exposeImplicitApps.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const t = require("babel-types");
const imm = require("immutable");
const helpers_1 = require("./helpers");
exports.implicitsIdentifier = t.identifier('$i');
const binopTable = imm.Map([
    ['+', 'add'],
    ['-', 'sub'],
    ['/', 'div'],
    ['*', 'mul']
]);
function implicit(name, ...args) {
    return t.callExpression(t.memberExpression(exports.implicitsIdentifier, t.identifier(name)), args);
}
const implicitsPath = `${helpers_1.runtimePath}/implicitApps`;
const visitor = {
    Program(path, state) {
        const opts = state.opts;
        if (!opts.requireRuntime) {
            return;
        }
        path.node.body.unshift(t.variableDeclaration('var', [t.variableDeclarator(exports.implicitsIdentifier, t.callExpression(t.identifier('require'), [t.stringLiteral(implicitsPath)]))]));
    },
    BinaryExpression(path) {
        const fun = binopTable.get(path.node.operator);
        if (typeof fun !== 'string') {
            return;
        }
        path.replaceWith(implicit(fun, path.node.left, path.node.right));
    },
    MemberExpression: {
        exit(path) {
            if (!path.node.computed) {
                return;
            }
            path.node.property = implicit('toKey', path.node.property);
        }
    }
};
function plugin() {
    return { visitor };
}
exports.plugin = plugin;
