// @stopify/continuations/dist/src/exposeGettersSetters.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const t = require("babel-types");
const helpers_1 = require("./helpers");
const gettersRuntime = t.identifier('$gs');
const get_prop = t.memberExpression(gettersRuntime, t.identifier('get_prop'));
const set_prop = t.memberExpression(gettersRuntime, t.identifier('set_prop'));
const delete_prop = t.memberExpression(gettersRuntime, t.identifier('delete_prop'));
// Stops $gs.get_prop and $gs.set_prop from causing an infinite loop.
set_prop.exposed = true;
get_prop.exposed = true;
delete_prop.exposed = true;
function $func(func, ...args) {
    return t.callExpression(func, args);
}
const gettersRuntimePath = `${helpers_1.runtimePath}/gettersSetters.js`;
let enableGetters = false;
const gettersUsedVisitor = {
    ObjectMethod(path) {
        if (path.node.kind !== 'method') {
            enableGetters = true;
            path.stop();
        }
    },
    MemberExpression(path) {
        const { object, property } = path.node;
        if (t.isIdentifier(object) && object.name === 'Object' &&
            t.isIdentifier(property) && property.name === 'defineProperty') {
            enableGetters = true;
            path.stop();
        }
    }
};
const visitor = {
    Program(path, state) {
        path.traverse(gettersUsedVisitor);
        if (!enableGetters) {
            console.log('No uses of getters found, disabling --getters');
            path.stop();
        }
        const opts = state.opts;
        if (!opts.requireRuntime) {
            return;
        }
        path.node.body.unshift(t.variableDeclaration('var', [t.variableDeclarator(gettersRuntime, t.callExpression(t.identifier('require'), [t.stringLiteral(gettersRuntimePath)]))]));
    },
    UnaryExpression: {
        enter(path) {
            const { operator, argument } = path.node;
            // Support property deletion.
            if (operator === 'delete' && t.isMemberExpression(argument)) {
                const { object, property, computed } = argument;
                const prop = computed ?
                    property :
                    t.stringLiteral(property.name);
                path.replaceWith($func(delete_prop, object, prop));
            }
        }
    },
    MemberExpression(path) {
        const p = path.parent;
        const { object, property } = path.node;
        // Only useful for console.log and various function on Math.
        if (path.node.mark === 'Flat') {
            return;
        }
        // Setters
        if (t.isAssignmentExpression(p) && p.left === path.node && p.operator === '=') {
            const prop = path.node.computed ?
                property :
                t.stringLiteral(property.name);
            path.parentPath.replaceWith($func(set_prop, object, prop, p.right));
        }
        else if (t.isUpdateExpression(p) || t.isAssignmentExpression(p) ||
            path.node.exposed) {
            return;
        }
        // Getters
        else {
            const prop = path.node.computed ?
                property :
                t.stringLiteral(property.name);
            path.replaceWith($func(get_prop, object, prop));
        }
    }
};
function plugin() {
    return { visitor };
}
exports.plugin = plugin;
