// @stopify/continuations/dist/src/index.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
const babel = require("babel-core");
const babylon = require("babylon");
const t = require("babel-types");
const callcc = require("./callcc/callcc");
const hygiene = require("@stopify/hygiene");
const h = require("@stopify/util");
const continuationsRTS = require("@stopify/continuations-runtime");
const cannotCapture_1 = require("./common/cannotCapture");
const exposeImplicitApps = require("./exposeImplicitApps");
const jumper_1 = require("./callcc/jumper");
var flatness_1 = require("./compiler/flatness");
exports.flatness = flatness_1.flatness;
var sourceMaps_1 = require("./compiler/sourceMaps");
exports.getSourceMap = sourceMaps_1.getSourceMap;
var callcc_1 = require("./callcc/callcc");
exports.plugin = callcc_1.default;
__export(require("./runtime/sentinels"));
var cannotCapture_2 = require("./common/cannotCapture");
exports.knownBuiltIns = cannotCapture_2.knownBuiltIns;
exports.reserved = [
    ...cannotCapture_1.knowns,
    "name",
    exposeImplicitApps.implicitsIdentifier.name,
    "$opts",
    "$result",
    "target",
    "newTarget",
    "captureLocals",
    jumper_1.restoreNextFrame.name,
    "frame",
    "RV_SENTINAL",
    "EXN_SENTINAL",
    "finally_rv",
    "finally_exn",
    "captureCC",
    'materializedArguments',
    'argsLen',
    '$top',
    '$S'
];
const visitor = {
    Program(path, state) {
        const opts = {
            getters: false,
            debug: false,
            captureMethod: 'lazy',
            newMethod: 'direct',
            es: 'sane',
            jsArgs: 'simple',
            requireRuntime: false,
            sourceMap: { getLine: () => null },
            // function done(r) { return $rts.onDone(r); }
            onDone: t.functionExpression(t.identifier('done'), [t.identifier('r')], t.blockStatement([
                t.returnStatement(t.callExpression(t.memberExpression(t.identifier('$rts'), t.identifier('onDone')), [t.identifier('r')]))
            ])),
            eval2: false,
            compileMode: 'normal'
        };
        h.transformFromAst(path, [[hygiene.plugin, state.opts]]);
        h.transformFromAst(path, [[callcc.default, opts]]);
        path.stop();
    }
};
/**
 * Compiles a program to support callCC.
 *
 * @param src the program to compile, which may use callCC
 * @returns an ordinary JavaScript program
 */
function compileFromAst(src) {
    try {
        const babelOpts = {
            plugins: [[() => ({ visitor }), {
                        reserved: [],
                        global: t.memberExpression(t.identifier('$rts'), t.identifier('g'))
                    }]],
            babelrc: false,
            ast: false,
            code: true,
            minified: false,
            comments: false,
        };
        const result = babel.transformFromAst(src, undefined, babelOpts);
        if (typeof result.code === 'string') {
            return h.ok(result.code);
        }
        else {
            return h.error('compile failed: no code returned');
        }
    }
    catch (exn) {
        return h.error(exn.toString());
    }
}
exports.compileFromAst = compileFromAst;
/**
 * Compiles a program to support callCC.
 *
 * @param src the program to compile, which may use callCC
 * @returns an ordinary JavaScript program
 */
function compile(src) {
    return h.asResult(() => babylon.parse(src).program)
        .then(p => compileFromAst(p))
        .map(code => new RunnerImpl(code));
}
exports.compile = compile;
class RunnerImpl {
    constructor(code) {
        this.code = code;
        this.g = Object.create(null);
    }
    run(onDone) {
        let stopify = continuationsRTS;
        this.rts = stopify.newRTS('lazy');
        let $rts = { g: this.g, onDone: onDone };
        return eval(this.code);
    }
    control(f) {
        return this.rts.captureCC(k => {
            return f((x) => k({ type: 'normal', value: x }));
        });
    }
    processEvent(body, receiver) {
        this.rts.runtime(body, receiver);
    }
}
