// @stopify/stopify/dist/src/stopify/compileFunction.js
"use strict";
/**
 * `func` compile mode should be used when function bodies need to be compiled
 * while preserving the function signatures. This is currently being used in
 * the pyret compiler.
 *
 * This passes around information to make sure that:
 * - the function signature is preserved
 * - globals are not redeclared (since the input function might capture variables)
 */
Object.defineProperty(exports, "__esModule", { value: true });
const babel = require("babel-core");
const t = require("babel-types");
const stopifyCallCC = require("./stopifyCallCC");
const assert = require("assert");
const util = require("@stopify/util");
var continuations_1 = require("@stopify/continuations");
exports.RV_SENTINAL = continuations_1.RV_SENTINAL;
exports.EXN_SENTINAL = continuations_1.EXN_SENTINAL;
exports.knownBuiltIns = continuations_1.knownBuiltIns;
const visitor = {
    Program: {
        enter(path, { opts }) {
            path.stop();
            assert.equal(path.node.body.length, 1);
            const func = path.node.body[0];
            if (func.type !== 'FunctionDeclaration') {
                throw new Error('Must compile a top-level function');
            }
            else {
                // If compile a string to be eval'd, convert last statement to a return
                // statement
                if (opts.eval) {
                    const lastStatement = func.body.body.pop();
                    if (lastStatement.type === 'ExpressionStatement') {
                        func.body.body.push(t.returnStatement(lastStatement.expression));
                    }
                    else {
                        func.body.body.push(lastStatement);
                    }
                }
            }
            util.transformFromAst(path, [[stopifyCallCC.plugin, opts]]);
        }
    }
};
const defaultOpts = {
    getters: false,
    compileFunction: true,
    debug: false,
    captureMethod: 'lazy',
    newMethod: 'wrapper',
    es: 'sane',
    jsArgs: 'simple',
    requireRuntime: false,
    onDone: t.functionExpression(t.identifier('onDone'), [], t.blockStatement([])),
    sourceMap: { getLine: (x, y) => null },
    eval2: false,
    compileMode: 'normal'
};
function compileFunction(code, opts = defaultOpts) {
    const babelOpts = {
        plugins: [[() => ({ visitor }), opts]],
        babelrc: false
    };
    const { code: transformed } = babel.transform(code, babelOpts);
    if (!transformed) {
        throw new Error("Failed to transform function");
    }
    return transformed;
}
exports.compileFunction = compileFunction;
function compileEval(code, compilerOpts, renames, boxes) {
    const toCompile = `function __eval__function() { ${code} }`;
    const transformed = compileFunction(toCompile, Object.assign({}, compilerOpts, { renames,
        boxes }));
    return `(${transformed})()`;
}
exports.compileEval = compileEval;
function default_1() {
    return { visitor };
}
exports.default = default_1;
