// @stopify/stopify/dist/src/stopify/suspendStop.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const t = require("babel-types");
function handleBlock(body) {
    body.body.unshift(t.expressionStatement(t.callExpression(t.memberExpression(t.identifier("$S"), t.identifier("suspend")), [])));
}
function handleFunction(path) {
    if (path.node.mark === 'Flat') {
        return;
    }
    handleBlock(path.node.body);
}
const insertSuspend = {
    FunctionDeclaration(path) {
        handleFunction(path);
    },
    FunctionExpression(path) {
        handleFunction(path);
    },
    Loop(path) {
        if (path.node.body.type === "BlockStatement") {
            handleBlock(path.node.body);
        }
        else {
            const body = t.blockStatement([path.node.body]);
            path.node.body = body;
            handleBlock(body);
        }
    },
    Program: {
        exit(path, { opts }) {
            if (opts.compileFunction && !opts.eval) {
                if (path.node.body[0].type === 'FunctionDeclaration') {
                    path.node.body[0].topFunction = true;
                }
                else {
                    throw new Error(`Compile function expected top-level functionDeclaration`);
                }
            }
        }
    },
};
function default_1() {
    return { visitor: insertSuspend };
}
exports.default = default_1;
