// @stopify/stopify/dist/src/runtime/suspend.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const setImmediate_1 = require("./setImmediate");
function badResume(x) {
    throw new Error('program is not paused. (Did you call .resume() twice?)');
}
exports.badResume = badResume;
function defaultDone(x) {
    if (x.type === 'exception') {
        console.error(x.value);
    }
}
exports.defaultDone = defaultDone;
const normalUndefResult = { type: 'normal', value: undefined };
/**
 * Instance of a runtime extended with the suspend() function. Used by
 * instrumented programs produced by stopify.
 */
class RuntimeWithSuspend {
    constructor(
    /**
     * Abstract runtime used to implement stack saving and restoring logic
     */
    rts, yieldInterval, estimator, 
    /** The runtime system yields control whenever this function produces
     * 'true' or when the estimated elapsed time exceeds 'yieldInterval'.
     */
    mayYield = function () { return false; }, 
    /** This function is applied immediately before stopify yields control to
     *  the browser's event loop. If the function produces 'false', the
     *  computation does not resume.
     */
    onYield = function () { return true; }, 
    /**
     * Called when execution reaches the end of any stopified module.
     */
    onEnd = function (x) { this.estimator.cancel(); defaultDone(x); }, continuation = badResume, onDone = defaultDone, 
    /**
     * Current line number in the source program. Used in `--debug` mode.
     */
    linenum = undefined) {
        this.rts = rts;
        this.yieldInterval = yieldInterval;
        this.estimator = estimator;
        this.mayYield = mayYield;
        this.onYield = onYield;
        this.onEnd = onEnd;
        this.continuation = continuation;
        this.onDone = onDone;
        this.linenum = linenum;
    }
    // Resume a suspended program.
    resumeFromCaptured() {
        const cont = this.continuation;
        const onDone = this.onDone;
        // Clear the saved continuation, or invoking .resumeFromCaptured() twice
        // in a row will restart the computation.
        this.continuation = badResume;
        this.onDone = defaultDone;
        return this.rts.runtime(() => cont(normalUndefResult), onDone);
    }
    /**
     * Call this function to suspend a running program. When called, it initiates
     * stack capturing by calling the `captureCC` function defined by the current
     * runtime.
     *
     * Internally uses stopify's timing mechanism to decide whether or not to
     * suspend.
     *
     * @param force forces a suspension when `true`.
     */
    suspend(force) {
        // If there are no more stack frame left to be consumed, save the stack
        // and continue running the program.
        if (isFinite(this.rts.stackSize) && this.rts.remainingStack <= 0) {
            this.rts.remainingStack = this.rts.stackSize;
            return this.rts.captureCC((continuation) => {
                if (this.onYield()) {
                    return continuation(normalUndefResult);
                }
            });
        }
        if (force || this.mayYield() ||
            (this.estimator.elapsedTime() >= this.yieldInterval)) {
            if (isFinite(this.rts.stackSize)) {
                this.rts.remainingStack = this.rts.stackSize;
            }
            this.estimator.reset();
            return this.rts.captureCC((continuation) => {
                return this.rts.endTurn((onDone) => {
                    this.continuation = continuation;
                    this.onDone = onDone;
                    if (this.onYield()) {
                        return setImmediate_1.setImmediate(() => {
                            this.rts.runtime(() => continuation(normalUndefResult), onDone);
                        });
                    }
                });
            });
        }
    }
}
exports.RuntimeWithSuspend = RuntimeWithSuspend;
