// @stopify/stopify/dist/src/entrypoints/compiler.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * This is the entrypoint for stopify-full.bundle.js. A page that includes
 * this entrypoint can compile programs in the browser.
 */
const babylon = require("babylon");
const abstractRunner_1 = require("../runtime/abstractRunner");
const compiler_1 = require("../compiler/compiler");
const check_compiler_opts_1 = require("@stopify/continuations/dist/src/compiler/check-compiler-opts");
const check_runtime_opts_1 = require("../runtime/check-runtime-opts");
const continuations_1 = require("@stopify/continuations");
// We need to provide these for stopify-continuations
__export(require("@stopify/continuations-runtime/dist/src/runtime/runtime"));
__export(require("@stopify/continuations/dist/src/runtime/implicitApps"));
var cannotCapture_1 = require("@stopify/continuations/dist/src/common/cannotCapture");
exports.knownBuiltIns = cannotCapture_1.knownBuiltIns;
const compiler = require("../stopify/compileFunction");
exports.compiler = compiler;
let runner;
function copyCompilerOpts(compileOpts) {
    return {
        compileFunction: compileOpts.compileFunction,
        getters: compileOpts.getters,
        debug: compileOpts.debug,
        captureMethod: compileOpts.captureMethod,
        newMethod: compileOpts.newMethod,
        es: compileOpts.es,
        jsArgs: compileOpts.jsArgs,
        requireRuntime: compileOpts.requireRuntime,
        sourceMap: compileOpts.sourceMap,
        onDone: compileOpts.onDone,
        eval2: compileOpts.eval2,
        compileMode: compileOpts.compileMode
    };
}
class Runner extends abstractRunner_1.AbstractRunner {
    constructor(code, compilerOpts, runtimeOpts) {
        super(runtimeOpts);
        this.code = code;
        this.evalOpts = copyCompilerOpts(compilerOpts);
        this.evalOpts.eval2 = true;
    }
    run(onDone, onYield, onBreakpoint) {
        this.runInit(onDone, onYield, onBreakpoint);
        eval.call(global, this.code);
    }
    evalAsyncFromAst(ast, onDone) {
        const stopifiedCode = compiler_1.compileFromAst(ast, this.evalOpts);
        this.eventMode = abstractRunner_1.EventProcessingMode.Running;
        this.continuationsRTS.runtime(eval(stopifiedCode), onDone);
    }
    evalAsync(src, onDone) {
        this.evalAsyncFromAst(babylon.parse(src).program, onDone);
    }
}
/**
 * Called by the stopified program to get suspend() and other functions.
 */
function init(rts) {
    if (runner === undefined) {
        throw new Error('stopify not called');
    }
    return runner.init(rts);
}
exports.init = init;
function stopifyLocallyFromAst(src, sourceMap, optionalCompileOpts, optionalRuntimeOpts) {
    try {
        const compileOpts = check_compiler_opts_1.checkAndFillCompilerOpts(optionalCompileOpts || {}, sourceMap);
        const runtimeOpts = check_runtime_opts_1.checkAndFillRuntimeOpts(optionalRuntimeOpts || {});
        const stopifiedCode = compiler_1.compileFromAst(src, compileOpts);
        runner = new Runner(stopifiedCode, compileOpts, runtimeOpts);
        return runner;
    }
    catch (exn) {
        return { kind: 'error', exception: exn };
    }
}
exports.stopifyLocallyFromAst = stopifyLocallyFromAst;
/**
 * Control the execution of a pre-compiled program.
 *
 * @param url URL of a pre-compiled program
 * @param opts runtime settings
 */
function stopifyLocally(src, optionalCompileOpts, optionalRuntimeOpts) {
    return stopifyLocallyFromAst(babylon.parse(src).program, continuations_1.getSourceMap(src), optionalCompileOpts, optionalRuntimeOpts);
}
exports.stopifyLocally = stopifyLocally;
