// @stopify/stopify/dist/src/entrypoints/runtimeOnly.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
const abstractRunner_1 = require("../runtime/abstractRunner");
const check_runtime_opts_1 = require("../runtime/check-runtime-opts");
// We need to provide these for stopify-continuations
__export(require("@stopify/continuations-runtime/dist/src/runtime/runtime"));
__export(require("@stopify/continuations/dist/src/runtime/implicitApps"));
// For testing / benchmarking convenience.
var parse_runtime_opts_1 = require("../parse-runtime-opts");
exports.parseRuntimeOpts = parse_runtime_opts_1.parseRuntimeOpts;
let runner;
class Runner extends abstractRunner_1.AbstractRunner {
    constructor(url, opts) {
        super(opts);
        this.url = url;
    }
    run(onDone, onYield, onBreakpoint) {
        this.runInit(onDone, onYield, onBreakpoint);
        const script = document.createElement('script');
        script.setAttribute('src', this.url);
        document.body.appendChild(script);
    }
}
/**
 * Called by the stopified program to get suspend() and other functions.
 */
function init(rts) {
    if (runner === undefined) {
        throw new Error('stopify not called');
    }
    return runner.init(rts);
}
exports.init = init;
/**
 * Control the execution of a pre-compiled program.
 *
 * @param url URL of a pre-compiled program
 * @param opts runtime settings
 */
function stopify(url, optionalRuntimeOpts) {
    const runtimeOpts = check_runtime_opts_1.checkAndFillRuntimeOpts(optionalRuntimeOpts || {});
    runner = new Runner(url, runtimeOpts);
    return runner;
}
exports.stopify = stopify;
